File: internal/dag/visualize.go
package dag

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Liam0205/pineapple/internal/types"
)

var dotColors = map[types.OperatorType]string{
	types.OpTypeRecall:    "#E8F5E9",
	types.OpTypeTransform: "#E3F2FD",
	types.OpTypeFilter:    "#FFF3E0",
	types.OpTypeMerge:     "#F3E5F5",
	types.OpTypeReorder:   "#FFFDE7",
	types.OpTypeObserve:   "#F5F5F5",
}

var mermaidClasses = map[types.OperatorType][2]string{
	types.OpTypeRecall:    {"#E8F5E9", "#4CAF50"},
	types.OpTypeTransform: {"#E3F2FD", "#2196F3"},
	types.OpTypeFilter:    {"#FFF3E0", "#FF9800"},
	types.OpTypeMerge:     {"#F3E5F5", "#9C27B0"},
	types.OpTypeReorder:   {"#FFFDE7", "#FFC107"},
	types.OpTypeObserve:   {"#F5F5F5", "#9E9E9E"},
}

// RenderDOT renders the DAG as a Graphviz DOT string.
func RenderDOT(g *Graph) string {
	var b strings.Builder
	b.WriteString("digraph pipeline {\n")
	b.WriteString("    rankdir=TB;\n")
	b.WriteString("    node [shape=box, style=filled, fontname=\"Helvetica\"];\n\n")

	for _, node := range g.Nodes {
		opType := types.OperatorType(node.Config.OperatorType)
		color := dotColors[opType]
		if color == "" {
			color = "#FFFFFF"
		}
		label := node.Name
		fmt.Fprintf(&b, "    %q [label=%q, fillcolor=%q];\n", node.Name, label, color)
	}

	b.WriteString("\n")

	for _, node := range g.Nodes {
		for _, succ := range node.Succs {
			fmt.Fprintf(&b, "    %q -> %q;\n", node.Name, g.Nodes[succ].Name)
		}
	}

	b.WriteString("}\n")
	return b.String()
}

// RenderMermaid renders the DAG as a Mermaid flowchart string.
func RenderMermaid(g *Graph) string {
	var b strings.Builder
	b.WriteString("graph TB\n")

	for _, node := range g.Nodes {
		opType := types.OperatorType(node.Config.OperatorType)
		className := strings.ToLower(string(opType))
		id := sanitizeMermaidID(node.Name)
		label := node.Name
		fmt.Fprintf(&b, "    %s[\"%s\"]:::%s\n", id, label, className)
	}

	b.WriteString("\n")

	for _, node := range g.Nodes {
		fromID := sanitizeMermaidID(node.Name)
		for _, succ := range node.Succs {
			toID := sanitizeMermaidID(g.Nodes[succ].Name)
			fmt.Fprintf(&b, "    %s --> %s\n", fromID, toID)
		}
	}

	b.WriteString("\n")

	for opType, colors := range mermaidClasses {
		className := strings.ToLower(string(opType))
		fmt.Fprintf(&b, "    classDef %s fill:%s,stroke:%s\n", className, colors[0], colors[1])
	}

	return b.String()
}

var mermaidIDRe = regexp.MustCompile(`[^a-zA-Z0-9_]`)

func sanitizeMermaidID(name string) string {
	return mermaidIDRe.ReplaceAllString(name, "_")
}

// collapsedGraph computes the aggregated view for level-based SubFlow rendering.
// Nodes are grouped by truncating their SubFlow path to the first `level`
// segments (split by "/"). Nodes with empty SubFlow remain independent.
type collapsedNode struct {
	Name  string // group key or original node name
	Group bool   // true if this represents a SubFlow group
}

type collapsedEdge struct {
	From, To int
}

// collapseKey returns the grouping key for a SubFlow path at a given level.
func collapseKey(subFlow string, level int) string {
	if subFlow == "" || level <= 0 {
		return ""
	}
	parts := strings.Split(subFlow, "/")
	if level >= len(parts) {
		return subFlow
	}
	return strings.Join(parts[:level], "/")
}

func buildCollapsed(g *Graph, level int) ([]collapsedNode, []collapsedEdge) {
	nodeToGroup := make(map[int]int)
	var groups []collapsedNode
	groupIndex := make(map[string]int) // key → group index

	for _, node := range g.Nodes {
		key := collapseKey(node.SubFlow, level)
		if key == "" {
			key = "\x00" + node.Name // unique per standalone node
		}
		if idx, ok := groupIndex[key]; ok {
			nodeToGroup[node.Index] = idx
		} else {
			idx := len(groups)
			if node.SubFlow != "" {
				groups = append(groups, collapsedNode{Name: collapseKey(node.SubFlow, level), Group: true})
			} else {
				groups = append(groups, collapsedNode{Name: node.Name, Group: false})
			}
			groupIndex[key] = idx
			nodeToGroup[node.Index] = idx
		}
	}

	edgeSet := make(map[[2]int]bool)
	var edges []collapsedEdge
	for _, node := range g.Nodes {
		fromG := nodeToGroup[node.Index]
		for _, succ := range node.Succs {
			toG := nodeToGroup[succ]
			if fromG == toG {
				continue // internal edge within same group
			}
			key := [2]int{fromG, toG}
			if !edgeSet[key] {
				edgeSet[key] = true
				edges = append(edges, collapsedEdge{From: fromG, To: toG})
			}
		}
	}

	return groups, edges
}

// RenderCollapsedDOT renders the DAG with SubFlow nodes collapsed into
// single aggregate nodes at the specified level.
func RenderCollapsedDOT(g *Graph, level int) string {
	groups, edges := buildCollapsed(g, level)
	var b strings.Builder
	b.WriteString("digraph pipeline {\n")
	b.WriteString("    rankdir=TB;\n")
	b.WriteString("    node [shape=box, style=filled, fontname=\"Helvetica\"];\n\n")

	for i, group := range groups {
		color := "#E0E0E0"
		if group.Group {
			color = "#BBDEFB"
		}
		fmt.Fprintf(&b, "    %q [label=%q, fillcolor=%q];\n",
			fmt.Sprintf("g%d", i), group.Name, color)
	}

	b.WriteString("\n")

	for _, e := range edges {
		fmt.Fprintf(&b, "    %q -> %q;\n",
			fmt.Sprintf("g%d", e.From), fmt.Sprintf("g%d", e.To))
	}

	b.WriteString("}\n")
	return b.String()
}

// RenderCollapsedMermaid renders the DAG with SubFlow nodes collapsed into
// single aggregate nodes at the specified level.
func RenderCollapsedMermaid(g *Graph, level int) string {
	groups, edges := buildCollapsed(g, level)
	var b strings.Builder
	b.WriteString("graph TB\n")

	for i, group := range groups {
		id := fmt.Sprintf("g%d", i)
		cls := "standalone"
		if group.Group {
			cls = "subflow"
		}
		fmt.Fprintf(&b, "    %s[\"%s\"]:::%s\n", id, group.Name, cls)
	}

	b.WriteString("\n")

	for _, e := range edges {
		fmt.Fprintf(&b, "    g%d --> g%d\n", e.From, e.To)
	}

	b.WriteString("\n")
	b.WriteString("    classDef subflow fill:#BBDEFB,stroke:#1976D2\n")
	b.WriteString("    classDef standalone fill:#E0E0E0,stroke:#616161\n")

	return b.String()
}

File: pine-python/pine/visualize.py
from __future__ import annotations

import re
from dataclasses import dataclass, field

from pine.dag import DAG

_DOT_FILL_COLORS = {
    "recall": "#E8F5E9",
    "transform": "#E3F2FD",
    "filter": "#FFF3E0",
    "merge": "#F3E5F5",
    "reorder": "#FFFDE7",
    "observe": "#F5F5F5",
}

_MERMAID_CLASS_DEFS = [
    "classDef recall fill:#E8F5E9,stroke:#4CAF50",
    "classDef transform fill:#E3F2FD,stroke:#2196F3",
    "classDef filter fill:#FFF3E0,stroke:#FF9800",
    "classDef merge fill:#F3E5F5,stroke:#9C27B0",
    "classDef reorder fill:#FFFDE7,stroke:#FFC107",
    "classDef observe fill:#F5F5F5,stroke:#9E9E9E",
]

_SANITIZE_RE = re.compile(r"[^a-zA-Z0-9_]")


def _sanitize_id(name: str) -> str:
    return _SANITIZE_RE.sub("_", name)


def render_dot(dag: DAG) -> str:
    lines: list[str] = []
    lines.append("digraph pipeline {")
    lines.append('    rankdir=TB;')
    lines.append('    node [shape=box, style=filled, fontname="Helvetica"];')
    lines.append("")

    for node in dag.nodes:
        color = _DOT_FILL_COLORS.get(node.config.operator_type, "#FFFFFF")
        lines.append(f'    "{node.name}" [label="{node.name}", fillcolor="{color}"];')

    lines.append("")

    for node in dag.nodes:
        for succ_idx in node.succs:
            succ = dag.nodes[succ_idx]
            lines.append(f'    "{node.name}" -> "{succ.name}";')

    lines.append("}")
    return "\n".join(lines) + "\n"


def render_mermaid(dag: DAG) -> str:
    lines: list[str] = []
    lines.append("graph TB")

    for node in dag.nodes:
        sid = _sanitize_id(node.name)
        op_type = node.config.operator_type or "transform"
        lines.append(f'    {sid}["{node.name}"]:::{op_type}')

    lines.append("")

    for node in dag.nodes:
        src_id = _sanitize_id(node.name)
        for succ_idx in node.succs:
            succ = dag.nodes[succ_idx]
            dst_id = _sanitize_id(succ.name)
            lines.append(f"    {src_id} --> {dst_id}")

    lines.append("")
    for cd in _MERMAID_CLASS_DEFS:
        lines.append(f"    {cd}")

    return "\n".join(lines) + "\n"


@dataclass
class _CollapsedGroup:
    key: str
    label: str
    is_subflow: bool
    node_indices: list[int] = field(default_factory=list)


def _collapse_key(sub_flow: str, level: int) -> str:
    if not sub_flow:
        return ""
    parts = sub_flow.split("/")
    if level >= len(parts):
        return sub_flow
    return "/".join(parts[:level])


def _build_collapsed(dag: DAG, level: int) -> tuple[list[_CollapsedGroup], list[tuple[int, int]]]:
    key_to_group_idx: dict[str, int] = {}
    groups: list[_CollapsedGroup] = []
    node_to_group: list[int] = []

    for node in dag.nodes:
        key = _collapse_key(node.sub_flow, level)
        if key == "":
            gidx = len(groups)
            groups.append(_CollapsedGroup(
                key=f"_standalone_{node.index}",
                label=node.name,
                is_subflow=False,
                node_indices=[node.index],
            ))
            node_to_group.append(gidx)
        elif key in key_to_group_idx:
            gidx = key_to_group_idx[key]
            groups[gidx].node_indices.append(node.index)
            node_to_group.append(gidx)
        else:
            gidx = len(groups)
            key_to_group_idx[key] = gidx
            groups.append(_CollapsedGroup(
                key=key,
                label=key,
                is_subflow=True,
                node_indices=[node.index],
            ))
            node_to_group.append(gidx)

    edges: list[tuple[int, int]] = []
    seen_edges: set[tuple[int, int]] = set()
    for node in dag.nodes:
        from_group = node_to_group[node.index]
        for succ_idx in node.succs:
            to_group = node_to_group[succ_idx]
            if from_group != to_group:
                edge = (from_group, to_group)
                if edge not in seen_edges:
                    seen_edges.add(edge)
                    edges.append(edge)

    return groups, edges


def render_collapsed_dot(dag: DAG, level: int) -> str:
    groups, edges = _build_collapsed(dag, level)

    lines: list[str] = []
    lines.append("digraph pipeline {")
    lines.append('    rankdir=TB;')
    lines.append('    node [shape=box, style=filled, fontname="Helvetica"];')
    lines.append("")

    for i, group in enumerate(groups):
        color = "#BBDEFB" if group.is_subflow else "#E0E0E0"
        lines.append(f'    "g{i}" [label="{group.label}", fillcolor="{color}"];')

    lines.append("")

    for from_g, to_g in edges:
        lines.append(f'    "g{from_g}" -> "g{to_g}";')

    lines.append("}")
    return "\n".join(lines) + "\n"


def render_collapsed_mermaid(dag: DAG, level: int) -> str:
    groups, edges = _build_collapsed(dag, level)

    lines: list[str] = []
    lines.append("graph TB")

    for i, group in enumerate(groups):
        cls = "subflow" if group.is_subflow else "standalone"
        lines.append(f'    g{i}["{group.label}"]:::{cls}')

    lines.append("")

    for from_g, to_g in edges:
        lines.append(f"    g{from_g} --> g{to_g}")

    lines.append("")
    lines.append("    classDef subflow fill:#BBDEFB,stroke:#1976D2")
    lines.append("    classDef standalone fill:#E0E0E0,stroke:#616161")

    return "\n".join(lines) + "\n"
